nengo/tests/test_synapses.py

```diff
@@ -137,38 +208,90 @@ def test_linearfilter_evaluate(plt):
 
 def test_linearfilter_y0(allclose):
     # --- y0 sets initial state correctly for high-order filter
-    synapse = LinearFilter(butter_num, butter_den, analog=False)
     v = 9.81
     x = v * np.ones(10)
-    assert allclose(synapse.filt(x, y0=v), v)
-    assert not allclose(synapse.filt(x, y0=0), v, record_rmse=False)
+    synapse = LinearFilter((butter_num, butter_den), analog=False, initial_output=v)
+    assert allclose(synapse.filt(x), v)
+    synapse = LinearFilter((butter_num, butter_den), analog=False, initial_output=0)
+    assert not allclose(synapse.filt(x), v, record_rmse=False, print_fail=0)
 
     # --- y0 does not work for high-order synapse when DC gain is zero
-    synapse = LinearFilter([1, 0], [1, 1])
+    synapse = LinearFilter(([1, 0], [1, 1]), initial_output=1)
     with pytest.raises(ValidationError, match="Cannot solve for state"):
-        synapse.filt(np.ones(10), y0=1)
+        synapse.filt(np.ones(10))
 
 
 def test_linearfilter_extras(allclose):
     # This filter is just a gain, but caused index errors previously
-    synapse = nengo.LinearFilter([3], [2])
+    synapse = nengo.LinearFilter(([3], [2]))
     assert allclose(synapse.filt([2.0]), 3)
 
     # differentiator should work properly
-    diff = nengo.LinearFilter([1, -1], [1, 0], analog=False)
-    assert allclose(diff.filt([1.0, -1.0, 2.0], y0=0), [1.0, -2.0, 3.0])
+    diff = nengo.LinearFilter(([1, -1], [1, 0]), analog=False)
+    assert allclose(diff.filt([1.0, -1.0, 2.0]), [1.0, -2.0, 3.0])
 
     # Filtering an integer array should cast to a float
     x = np.arange(10, dtype=nengo.rc.int_dtype)
-    synapse = nengo.LinearFilter([1], [0.005, 1])
+    synapse = nengo.LinearFilter(([1], [0.005, 1]))
     assert synapse.filt(x).dtype == nengo.rc.float_dtype
 
     # Throw an error if non-float dtype
     shape = (1,)
     with pytest.raises(ValidationError, match="Only float data types"):
-        synapse.make_state(shape, shape, dt=0.001, dtype=np.int32)
+        synapse.make_state(shape, shape, dt=0.001, rng=None, dtype=np.int32)
     with pytest.raises(ValidationError, match="Only float data types"):
-        synapse.make_state(shape, shape, dt=0.001, dtype=np.complex64)
+        synapse.make_state(shape, shape, dt=0.001, rng=None, dtype=np.complex64)
+
+
+def test_linearfilter_den():
+    sys = ([1], [0.005, 1])
+    with pytest.warns(DeprecationWarning, match="`den` is deprecated"):
+        synapse = nengo.LinearFilter(*sys)
+
+    assert synapse == nengo.LinearFilter(sys)
+
+
+@pytest.mark.parametrize(
+    "synapse, rmse_range",
+    [
+        (DiscreteDelay(20), [0, 1e-8]),
+        (LegendreDelay(0.09, 9), [0.001, 0.03]),
+        (LegendreDelay(0.09, 21), [0.001, 0.03]),
```

Comment:
I think something might be off by +/- 1 time-steps as this tolerance should be able to be dropped significantly for larger orders. For example, if I bump the order up to 512 then the unit test generates thie plot:

![Screenshot from 2020-08-30 15-35-23](https://user-images.githubusercontent.com/5420057/91667981-86947a80-ead6-11ea-8af8-ea464924bc31.png)

which looks like it might be off by one step (or maybe [half a step](https://github.com/arvoelke/nengolib/issues/169)). The RMSE should be able to approach 0 as this is scaled up.

I spent some time fiddling with a few things in the unit test and couldn't figure it out though. So unless you can think of something I think it's fine as is (they plots look pretty accurate).

Reply:
It does look slightly off. I'm not sure about the reason either, though.